browser/extensions/manifest_v2/brave_extensions_manifest_v2_migrator.cc

```diff
@@ -0,0 +1,425 @@
+
+  base::ScopedClosureRunner clear_settings_on_error(base::BindOnce(
+      [](const extensions::ExtensionId& brave_hosted_extension_id,
+         const base::FilePath& profile_dir) {
+        ClearExtensionSettingsOnFileThread(brave_hosted_extension_id,
+                                           profile_dir);
+      },
+      brave_hosted_extension_id, profile_dir));
+
+  if (!ClearExtensionSettingsOnFileThread(brave_hosted_extension_id,
+                                          profile_dir)) {
+    return;
+  }
+
+  const auto local_settings_backup =
+      GetLocalSettingsForImport(brave_hosted_extension_id, profile_dir);
+  if (!local_settings_backup.empty()) {
+    if (!base::Move(
+            local_settings_backup,
+            profile_dir.Append(extensions::kLocalExtensionSettingsDirectoryName)
+                .AppendASCII(brave_hosted_extension_id))) {
+      return;
+    }
+  }
+
+  bool error = false;
+  GetIndexedSettingsForImport(brave_hosted_extension_id, profile_dir)
+      .ForEach([&error, &profile_dir, &brave_hosted_extension_id,
+                &webstore_extension_id](const base::FilePath& path) {
+        if (error) {
+          return;
+        }
+        auto name = path.BaseName().value();
+        base::ReplaceFirstSubstringAfterOffset(
+            &name, 0, ASCIIToPathStringType(webstore_extension_id),
+            ASCIIToPathStringType(brave_hosted_extension_id));
+        if (!base::Move(path, profile_dir.Append(kIndexedDBDir).Append(name))) {
+          error = true;
+        }
+      });
+  if (!error) {
+    clear_settings_on_error.ReplaceClosure(base::DoNothing());
+  }
+}
+
+}  // namespace
+
+namespace extensions_mv2 {
+
+ExtensionsManifestV2Migrator::ExtensionsManifestV2Migrator(Profile* profile)
+    : profile_(profile) {
+  auto* registry = extensions::ExtensionRegistry::Get(profile_);
+  auto* extension_prefs = extensions::ExtensionPrefs::Get(profile_);
+
+  prefs_observation_.Observe(extension_prefs);
+  registry_observation_.Observe(registry);
+
+  for (const auto webstore_extension : kWebStoreHosted) {
+    const auto disable_reasons = extension_prefs->GetDisableReasons(
+        extensions::ExtensionId(webstore_extension.first));
+    if (disable_reasons.contains(
+            extensions::disable_reason::DISABLE_UNSUPPORTED_MANIFEST_VERSION)) {
+      OnExtensionDisableReasonsChanged(
+          extensions::ExtensionId(webstore_extension.first), disable_reasons);
+    }
+  }
+}
+
+ExtensionsManifestV2Migrator::~ExtensionsManifestV2Migrator() {
+  CHECK(!prefs_observation_.IsObserving());
+  CHECK(!registry_observation_.IsObserving());
+}
+
+void ExtensionsManifestV2Migrator::Shutdown() {
+  prefs_observation_.Reset();
+  registry_observation_.Reset();
+}
+
+void ExtensionsManifestV2Migrator::OnExtensionPrefsWillBeDestroyed(
+    extensions::ExtensionPrefs* prefs) {
+  prefs_observation_.Reset();
+}
+
+void ExtensionsManifestV2Migrator::OnExtensionDisableReasonsChanged(
```

Comment:
How it suppose to work with sync enabled?
If a user click `Remove` on one machine, does it result in just dropping extension on another one (without `OnExtensionDisableReasonsChanged` triggering)?